MaxText/maxengine.py

```diff
@@ -796,11 +826,17 @@ def generate(
       rng: Optional[PRNGKeyType] = None,
   ) -> Tuple[DecodeState, engine_api.ResultTokens]:
     """Public API for generate that updates page state outside JIT."""
-    # Update page state before JIT call
 
+    # Update page state before JIT call
     if self.page_manager is not None and self.page_state is not None:
       self.page_state = self.page_manager.update_decode_pages(self.page_state)
 
+    # Sample rng before JIT call
+    if rng is None:
+      if self.rng is None:
+        self.rng = jax.random.PRNGKey(0)
+      self.rng, rng = jax.random.split(self.rng)
```

Comment:
Nit: It's fine to submit as-is, but this could be a separate function since it looks like it's here 3 times